src/coreclr/src/vm/proftoeeinterfaceimpl.cpp

```diff
@@ -7047,22 +7226,91 @@ HRESULT ProfToEEInterfaceImpl::EventPipeCreateProvider(const WCHAR *szName, EVEN
         "**PROF: EventPipeCreateProvider.\n"));
 
 #ifdef FEATURE_PERFTRACING
-    if (szName == NULL || pProviderHandle == NULL)
+    if (szName == NULL || pProvider == NULL)
     {
         return E_INVALIDARG;
     }
 
     HRESULT hr = S_OK;
     EX_TRY
     {
-        EventPipeProvider *pProvider = EventPipe::CreateProvider(szName, NULL, NULL);
-        if (pProvider == NULL)
+        EventPipeProvider *pRealProvider = EventPipe::CreateProvider(szName, NULL, NULL);
+        if (pRealProvider == NULL)
         {
             hr = E_FAIL;
         }
         else
         {
-            *pProviderHandle = reinterpret_cast<EVENTPIPE_PROVIDER>(pProvider);
+            *pProvider = reinterpret_cast<EVENTPIPE_PROVIDER>(pRealProvider);
+        }
+    }
+    EX_CATCH_HRESULT(hr);
+
+    return hr;
+#else // FEATURE_PERFTRACING
+    return E_NOTIMPL;
+#endif // FEATURE_PERFTRACING
+}
+
+HRESULT ProfToEEInterfaceImpl::EventPipeGetProviderInfo(
+            EVENTPIPE_PROVIDER provider,
+            ULONG      cchName,
+            ULONG      *pcchName,
+            WCHAR      szName[])
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+        EE_THREAD_NOT_REQUIRED;
+    }
+    CONTRACTL_END;
+
+    PROFILER_TO_CLR_ENTRYPOINT_ASYNC_EX(kP2EEAllowableAfterAttach,
+        (LF_CORPROF,
+        LL_INFO1000,
+        "**PROF: EventPipeGetProviderInfo.\n"));
+
+#ifdef FEATURE_PERFTRACING
+        if (cchName == 0 || szName == NULL)
+        {
+            // The only thing this API returns is the name, so bail
+            // if we can't return the name.
+            return E_INVALIDARG;
+        }
+
+        EventPipeProvider *pRealProvider = reinterpret_cast<EventPipeProvider *>(provider);
+        if (pRealProvider == NULL)
+        {
+            // Bogus provider passed in
+            return E_INVALIDARG;
+        }
+
+    HRESULT hr = S_OK;
+    EX_TRY
+    {
+        const SString &providerName = pRealProvider->GetProviderName();
```

Comment:
So we need the string allocation here? Any way to avoid it?

Reply:
I don't think this actually does an allocation, we already have the provider name on the EventPipeProvider object. And I wouldn't expect this API to be on any hot paths, you would only call this if you want to figure out the name. You would generally use the handle to identify the provider you care about and only call this if you want to do some formatting. There's no reason you couldn't cache the result